[TypeChecker] SE-0352: Require coercion if result type contains existential(s) that would loose generic requirements

xedin opened the pull request:
Implements SE-0352 revision - require explicit `as` coercion when existential
erasure loses information. 

For example:

```swift
protocol P {
  associatedtype A
}

protocol Q {
  associatedtype B: P where B.A == Int
}

func getB<T: Q>(_: T) -> T.B { ... }

func test(v: any Q) {
  let _ = getB(v) // <- produces `any P` which loses A == Int
}
```

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSFix] Add a skeleton fix to require explicit existential coercion

xedin committed:
[Diagnostics] Diagnose returning existential values that lose requirements

Diagnose situations where inferring existential type for result of
a call would result in loss of generic requirements.

```swift
protocol P {
  associatedtype A
}

protocol Q {
 associatedtype B: P where B.A == Int
}

func getB<T: Q>(_: T) -> T.B { ... }

func test(v: any Q) {
  let _ = getB(v) // <- produces `any P` which looses A == Int
}
```

xedin committed:
[CSFix] Implement missing existential coercion fix

xedin committed:
[CSSimplify] Start checking whether result type of a call needs a coercion

If there is going to be any loss of generic requirements in the result,
a result of a call has to use explicit coercion.

xedin commented:
@DougGregor Please take a look to make sure that this is going in the right direction, I still need to implement this diagnostic for Self erasure cases but it shouldn't be too bad.

xedin commented on lib/Sema/CSFix.cpp:
```diff
@@ -2151,3 +2151,127 @@ IgnoreDefaultExprTypeMismatch::create(ConstraintSystem &cs, Type argType,
+    }
+
+  private:
+    bool hasConstrainedAssociatedTypes(ArchetypeType *archetypeTy) {
```
I realized that this condition is incorrect, I need to check whether any member in the chain of associated types described by dependent member ends up having a where clause.

xedin replied:
Fixed!

xedin committed:
[CSFix] Refine coercion check for dependent member chains

For types like `<Base>.B.C` we need to check that neither
`B` nor `C` have any additional requirements because they
cannot be expressed in the existential type.

xedin committed:
[TypeChecker] Require a coercion if result of protocol member access would loose information

Accessing members on the protocol could result in existential opening and subsequence
result erasure, which requires explicit coercion if there is any loss of generic requirements.

xedin committed:
[Diagnostics] Attach note about erased result type to the call instead of declaration

Since the inference is a call site specific it makes sense to attach
a candidate note to the call itself instead of the declaration used.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke test

DougGregor commented on lib/Sema/CSFix.cpp:
```diff
@@ -2151,3 +2152,162 @@ IgnoreDefaultExprTypeMismatch::create(ConstraintSystem &cs, Type argType,
+      auto *assocType = member->getAssocType();
+
+      // If this member has associated type requirements, we are done.
+      if (assocType->getTrailingWhereClause())
```
This is looking at the syntactic construct (the trailing where clause), which is not serialized and may not have complete information. Can we instead look at the generic signature to see if there are any requirements that mention this dependent member type?

xedin replied:
Ah, I didn't know that this was the case, will fix it before merging this.

DougGregor commented on lib/Sema/CSFix.cpp:
```diff
@@ -2151,3 +2152,162 @@ IgnoreDefaultExprTypeMismatch::create(ConstraintSystem &cs, Type argType,
+      for (auto *protocol : layout.getProtocols()) {
+        if (llvm::any_of(protocol->getAssociatedTypeMembers(),
+                         [](const auto *assocTypeDecl) {
+                           return bool(assocTypeDecl->getTrailingWhereClause());
```
Same concern as above: `getTrailingWhereClause` is more syntactic than semantic.

DougGregor commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -8143,3 +8143,24 @@ bool DefaultExprTypeMismatch::diagnoseAsError() {
+    InFlightDiagnostic &diagnostic) const {
+  auto printOpts = PrintOptions::forDiagnosticArguments();
+  diagnostic.fixItInsertAfter(getSourceRange().End,
+                              "as " + ErasedResultType->getString(printOpts));
```
Could we insert parentheses in cases where the expression being coerced is an argument to a function that can be opened?

xedin replied:
Sure!

xedin replied:
Implemented!

xedin committed:
[CSFix] Check for constrained associated types via generic signature

Instead of checking for presence of `where` clause, let's use
requirement signature and check whether any of the requires are
anchored on a particular associated type.

xedin committed:
[Diagnostics] Prevent fix-it for explicit existential erasure from suppressing opening at use site

If erased result is passed as an argument to a call that requires
implicit opening, the fix-it should use parens to avoid suppressing
the opening at that argument position.

xedin committed:
[CSSimplify] Don't suppress existential opening if `as any ...` is in parens

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test macOS platform
